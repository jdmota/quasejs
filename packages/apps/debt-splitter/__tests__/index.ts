import { expect, it } from "@jest/globals";
import {
  computeGraph,
  cost,
  cover,
  everyone,
  parseCurrency,
  printCurrency,
  tableToString,
  type CostsRegistry,
} from "../debt-split";

it("valid currency", () => {
  expect(parseCurrency("-24.00")).toBe(-2400n);
  expect(parseCurrency("24.00")).toBe(2400n);
  expect(parseCurrency("24.0")).toBe(2400n);
  expect(parseCurrency("24")).toBe(2400n);
  expect(parseCurrency(".00")).toBe(0n);
  expect(parseCurrency(".10")).toBe(10n);
  expect(parseCurrency(".1")).toBe(10n);
});

it("invalid currency", () => {
  ["24.", "24..", "24.000", "-"].forEach(c => {
    expect(() => parseCurrency(c)).toThrow(`Invalid value: ${c}`);
  });
});

it("debt split with hero strat", () => {
  const carA = ["A", "E", "L", "L2"];
  const carJ2 = ["A2", "I", "J", "J2"];
  const J = ["J"];
  const L2 = ["L2"];
  const jAndJ2 = ["J", "J2"];

  const costs: CostsRegistry = {
    A: [cost("24.66", "", everyone), cost("22.00", "", carA)],
    A2: [
      cost("10.00", "", everyone),
      cost("1.3", "", everyone),
      cost("28", "", everyone),
      cost("10", "", everyone),
      cost("8.67", "", J),
      cost("2.00", "", carJ2),
    ],
    E: [cost("54.39", "", everyone), cost("11.50", "", L2)],
    I: [],
    J: [
      cost("96.52", "", everyone, [cover("2.16", "E", "")]),
      cost("9.66", "", everyone, [cover("0.39", "E", "")]),
    ],
    J2: [
      cost("2.00", "", jAndJ2),
      cost("27.56", "", carJ2),
      cost("17.30", "", carJ2),
      cost("55.00", "", carJ2),
    ],
    L: [
      cost("25.95", "", everyone),
      cost("1.26", "", everyone),
      cost("8.29", "", everyone),
      cost("-3.90", "", everyone),
      cost("-1.50", "", everyone),
      cost("-1.40", "", everyone),
      cost("-1.45", "", everyone),
    ],
    L2: [
      cost("35.72", "", everyone),
      cost("58.88", "", everyone, [cover("2.19", "E", "")]),
      cost("5.20", "", everyone),
      cost("8.95", "", J),
    ],
  };

  const {
    graph,
    net,
    originalCost,
    diff,
    sanityCheck,
    givers,
    receivers,
    transactions,
    leftOversExisted,
  } = computeGraph(costs, "J");

  expect(tableToString(graph)).toMatchSnapshot("Graph");
  expect(printCurrency(originalCost)).toMatchSnapshot("Original Total Cost");
  expect(tableToString(diff)).toMatchSnapshot("Diff");
  expect(sanityCheck).toBe(0n);
  expect(tableToString(net)).toMatchSnapshot("Net");
  expect(tableToString(givers)).toMatchSnapshot("Givers");
  expect(tableToString(receivers)).toMatchSnapshot("Receivers");
  expect(tableToString(transactions)).toMatchSnapshot("Transactions");
  expect(leftOversExisted).toBe(true);
});
