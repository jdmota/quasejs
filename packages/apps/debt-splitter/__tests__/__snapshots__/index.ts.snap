// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`debt split with hero strat: Diff 1`] = `
{
  "A": "0.00",
  "A2": "0.00",
  "E": "0.00",
  "I": "0.00",
  "J": "0.00",
  "J2": "0.00",
  "L": "0.00",
  "L2": "0.00",
}
`;

exports[`debt split with hero strat: Givers 1`] = `
{
  "A": "-3.42",
  "A2": "-10.07",
  "I": "-70.04",
  "L": "-22.83",
}
`;

exports[`debt split with hero strat: Graph 1`] = `
{
  "A": {
    "A": "8.58",
    "A2": "6.16",
    "E": "6.79",
    "I": "0.00",
    "J": "12.95",
    "J2": "0.00",
    "L": "3.40",
    "L2": "12.20",
  },
  "A2": {
    "A": "3.08",
    "A2": "6.66",
    "E": "6.79",
    "I": "0.00",
    "J": "12.95",
    "J2": "24.96",
    "L": "3.40",
    "L2": "12.20",
  },
  "E": {
    "A": "8.58",
    "A2": "6.16",
    "E": "6.79",
    "I": "0.00",
    "J": "15.50",
    "J2": "0.00",
    "L": "3.40",
    "L2": "14.39",
  },
  "I": {
    "A": "3.08",
    "A2": "6.66",
    "E": "6.79",
    "I": "0.00",
    "J": "12.95",
    "J2": "24.96",
    "L": "3.40",
    "L2": "12.20",
  },
  "J": {
    "A": "3.10",
    "A2": "15.35",
    "E": "6.86",
    "I": "0.00",
    "J": "12.98",
    "J2": "25.98",
    "L": "3.45",
    "L2": "21.16",
  },
  "J2": {
    "A": "3.08",
    "A2": "6.66",
    "E": "6.79",
    "I": "0.00",
    "J": "12.95",
    "J2": "25.96",
    "L": "3.40",
    "L2": "12.20",
  },
  "L": {
    "A": "8.58",
    "A2": "6.16",
    "E": "6.79",
    "I": "0.00",
    "J": "12.95",
    "J2": "0.00",
    "L": "3.40",
    "L2": "12.20",
  },
  "L2": {
    "A": "8.58",
    "A2": "6.16",
    "E": "18.29",
    "I": "0.00",
    "J": "12.95",
    "J2": "0.00",
    "L": "3.40",
    "L2": "12.20",
  },
}
`;

exports[`debt split with hero strat: Net 1`] = `
{
  "A": "-3.42",
  "A2": "-10.07",
  "E": "11.07",
  "I": "-70.04",
  "J": "17.30",
  "J2": "30.82",
  "L": "-22.83",
  "L2": "47.17",
}
`;

exports[`debt split with hero strat: Original Total Cost 1`] = `"516.56"`;

exports[`debt split with hero strat: Receivers 1`] = `
{
  "E": "11.07",
  "J": "17.30",
  "J2": "30.82",
  "L2": "47.17",
}
`;

exports[`debt split with hero strat: Transactions 1`] = `
{
  "A": {
    "E": "1.00",
    "J": "0.00",
    "J2": "2.42",
    "L2": "0.00",
  },
  "A2": {
    "E": "10.07",
    "J": "0.00",
    "J2": "0.00",
    "L2": "0.00",
  },
  "I": {
    "E": "0.00",
    "J": "0.00",
    "J2": "22.87",
    "L2": "47.17",
  },
  "L": {
    "E": "0.00",
    "J": "17.30",
    "J2": "5.53",
    "L2": "0.00",
  },
}
`;
